smart_open/tests/test_smart_open.py

```diff
@@ -200,93 +202,158 @@ def test_http_pass(self):
-        # decompress the gzip and get the same md5 hash
-        self.assertEqual(m.hexdigest(), expected_hash)
+        self._test_compressed_http(".gz", False)
 
-    @responses.activate
     def test_http_bz2(self):
-        """Can open bz2 via http?"""
-        test_string = b'Hello World Compressed.'
-        #
-        # TODO: why are these tests writing to temporary files?  We can do the
-        # bz2 compression in memory.
-        #
-        with tempfile.NamedTemporaryFile('wb', suffix='.bz2', delete=False) as infile:
-            test_file = infile.name
+        """Can open bzip2 via http?"""
+        self._test_compressed_http(".bz2", False)
 
-        with smart_open.smart_open(test_file, 'wb') as outfile:
-            outfile.write(test_string)
-
-        with open(test_file, 'rb') as infile:
-            compressed_data = infile.read()
-
-        if os.path.isfile(test_file):
-            os.unlink(test_file)
-
-        responses.add(responses.GET, "http://127.0.0.1/data.bz2", body=compressed_data, stream=True)
-        smart_open_object = smart_open.smart_open("http://127.0.0.1/data.bz2")
-
-        # decompress the bzip2 and get the same md5 hash
-        self.assertEqual(smart_open_object.read(), test_string)
-
-    @responses.activate
     def test_http_xz(self):
         """Can open xz via http?"""
-        test_string = b'Hello World Compressed.'
-        #
-        # TODO: why are these tests writing to temporary files?  We can do the
-        # lzma compression in memory.
-        #
-        with tempfile.NamedTemporaryFile('wb', suffix='.xz', delete=False) as infile:
-            test_file = infile.name
+        self._test_compressed_http(".xz", False)
 
-        with smart_open.smart_open(test_file, 'wb') as outfile:
-            outfile.write(test_string)
+    @unittest.skipIf(six.PY2, 'gzip support for Py2 is not implemented yet')
+    def test_http_gz_query(self):
+        """Can open gzip via http with a query appended to URI?"""
+        self._test_compressed_http(".gz", True)
 
-        with open(test_file, 'rb') as infile:
-            compressed_data = infile.read()
+    def test_http_bz2_query(self):
+        """Can open bzip2 via http with a query appended to URI?"""
+        self._test_compressed_http(".bz2", True)
 
-        if os.path.isfile(test_file):
-            os.unlink(test_file)
+    def test_http_xz_query(self):
+        """Can open xz via http with a query appended to URI?"""
+        self._test_compressed_http(".xz", True)
 
-        responses.add(responses.GET, "http://127.0.0.1/data.xz", body=compressed_data, stream=True)
-        smart_open_object = smart_open.smart_open("http://127.0.0.1/data.xz")
 
-        # decompress the xz and get the same md5 hash
-        self.assertEqual(smart_open_object.read(), test_string)
+class SmartOpenFileObjTest(unittest.TestCase):
+    """
+    Test passing raw file objects.
+    """
+    text = 'Hello, world!'
 
+    def patch_string_io(func):
```

Comment:
First, this does not need to be a static method. It's simpler as a pure function.

Also, I think this patch_string_io + decorator stuff is more complicated than it needs to be. AFAIK, it's there because you want the buffer to be with-able (a context manager). Wouldn't this be simpler?

```python
def make_named_buffer(klass=six.BytesIO, value=None, name=None):
    buf = klass(value) if value else klass()
    if name:
        buf.name = name
    if six.PY2:
        buf.__enter__ = lambda self: self
        buf.__exit__ = lambda self, exc_type, exc_val, exc_tb: None
    return buf

binbuf = make_named_buffer(name='binary')
strbuf = make_named_buffer(six.StringIO, name='text')
```

That hits several birds with one stone:

- Shortcuts creating a named buffer (either binary or text) - see [previous review](https://github.com/RaRe-Technologies/smart_open/pull/263#discussion_r259609156)
- Simplifies the existing code (no wrappers, functools, decorators, static methods or other voodoo)

Reply:
Funnily enough, this was the previous code before I decided to go decorated. Changing.

Reply:
Python 2 fails with this code

```
======================================================================
ERROR: test_read_bytes (__main__.SmartOpenFileObjTest)
Can we read bytes from a byte stream?
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_smart_open.py", line 278, in test_read_bytes
    with smart_open.smart_open(buffer, 'rb') as sf:
TypeError: <lambda>() takes exactly 1 argument (0 given)

======================================================================
ERROR: test_write_bytes (__main__.SmartOpenFileObjTest)
Can we write bytes to a byte stream?
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_smart_open.py", line 285, in test_write_bytes
    with smart_open.smart_open(buffer, 'wb') as sf:
TypeError: <lambda>() takes exactly 1 argument (0 given)

----------------------------------------------------------------------
```

Reply:
I see, because we set the attribute on the instance, not on the type